### ISSUE TITLE
CUDNN batchnorm backprop doesn't work properly in evaluation mode

### ISSUE THREAD

**Issue by tgeorgy** [pytorch/pytorch#586]
Gradient changes, though inputs, outputs, and internal state stay same.
```python
cudnn.enabled = True

net = nn.BatchNorm1d(3).cuda()
net.eval()

idat = Variable(torch.rand(4,3).cuda(), requires_grad=True)
grad = torch.rand(4,3).cuda()

# 1st pass
res = net(idat)
res.backward(grad)
grad0 = idat.grad.data.cpu()

# 2nd pass
idat.grad.data.zero_()

res = net(idat)
res.backward(grad)
grad1 = idat.grad.data.cpu()

print (grad1 == grad0).all()
```
Even when I run forward in evaluation mode and backward in training I still get the gradient changed.
It works fine with cudnn disabled.

**Comment by szagoruyko** [comment]
cudnn does not support backprop in evaluate mode, we should add a fallback to nn for this case

**Comment by fmassa** [comment]
Hi,

thanks for the report. It turns out that the cudnn library (at least up to v5) doesn't support backward in evaluation mode.
We should assert that in the code.

**Comment by tgeorgy** [comment]
Hi, thanks for the responses.

What about another case when you run forward in evaluation mode and backward in training?
It produces same issue, but seems to work fine in lua torch.
Here is the code with changes:
```python
cudnn.enabled = True

net = nn.BatchNorm1d(3).cuda()

idat = Variable(torch.rand(4,3).cuda(), requires_grad=True)
grad = torch.rand(4,3).cuda()

# 1st pass
net.eval()
res = net(idat)
net.train()
res.backward(grad)
grad0 = idat.grad.data.cpu()

# 2nd pass
idat.grad.data.zero_()

net.eval()
res = net(idat)
net.train()
res.backward(grad)
grad1 = idat.grad.data.cpu()

assert (grad1 == grad0).all(), 'Grads have to be same'
```
If there is another way to implement virtual batch normalization - please share.

**Comment by soumith** [comment]
this should be fixed in master. It will be incorporated into the next release that will have binaries late today.

**Comment by ngimel** [comment]
Out of curiosity, why do backward in evaluate mode? (Or rather, what is virtual batch normalization?)

**Comment by soumith** [comment]
@ngimel a lot of research involves this. For example neural style transfer, or [adversarial examples](https://arxiv.org/abs/1412.6572) etc.
